sdk-typed-search-attributes.md

```diff
@@ -0,0 +1,497 @@
+func (SearchAttributeTime) Delete(Context)         { panic("TODO") }
+
+type SearchAttributeKeywordList string
+func (SearchAttributeKeywordList) Exists(Context) bool   { panic("TODO") }
+func (SearchAttributeKeywordList) Get(Context) []string  { panic("TODO") }
+func (SearchAttributeKeywordList) Set(Context, []string) { panic("TODO") }
+func (SearchAttributeKeywordList) Delete(Context)        { panic("TODO") }
+```
+
+Which can be defined and used like:
+
+```go
+const MySearchAttribute = SearchAttributeInt("MySearchAttribute")
+```
+
+And then used from the workflow. Alternatively we could have something like:
+
+```go
+type searchAttributeValueConstraint interface {
+	string | int | float64 | time.Time | []string
+}
+
+type SearchAttribute[T searchAttributeValueConstraint] string
+
+func (SearchAttribute[T]) Exists(Context) bool { panic("TODO") }
+func (SearchAttribute[T]) Get(Context) T       { panic("TODO") }
+func (SearchAttribute[T]) Set(Context, T)      { panic("TODO") }
+func (SearchAttribute[T]) Delete(Context)      { panic("TODO") }
+```
+
+And defined like:
+
+```go
+const MySearchAttribute = SearchAttribute[int]("MySearchAttribute")
+```
+
+But the former is more explicit probably. This same thing would need to be done for memos but be looser on accepted
+types.
+
+### Java
+
+Visibility and method bodies left out for brevity.
+
+```java
+public class SearchAttributeMap implements Map<String, Object> {
+    // Constructor fails if called in workflow
+    SearchAttributeMap()
+
+    @Deprecated
+    Object get(Object key);
+    @Deprecated
+    Object put(String key, Object value);
+
+    String getText(String key);
+    String putText(String key, String value);
+    String getKeyword(String key);
+    String putKeyword(String key, String value);
+    Integer getInteger(String key);
+    String putInteger(String key, Integer value);
+    Double getDouble(String key);
+    String putDouble(String key, Double value);
+    OffsetDateTime getOffsetDateTime(String key);
+    OffsetDateTime putDouble(String key, OffsetDateTime value);
+    // Immutable
+    List<String> getKeywordList(String key);
+    List<String> putKeywordList(String key, List<String> value);
+}
+```
+
+* Deprecate `WorkflowOptions.Builder.setSearchAttributes`
+* Add overload `WorkflowOptions.Builder.setSearchAttributes` that accepts this map type
+  * The ability to need to instantiate this is why it's a class and not an interface
+    * Could have this be an interface if it was also easy to create somehow
+* Deprecate `Workflow.getSearchAttributes`, `Workflow.getSearchAttribute`, and `Workflow.getSearchAttributeValues`
+  * Can't think of a better way here, we can't reuse these from my POV and still be compat
+* Add `Workflow.getSearchAttributeMap` that returns this
+  * This is mutable! Mutations follow general rules laid out before
+* Should we log a runtime warning if they use the Map accessors directly?
```

Comment:
Same Q here - why implement `Map` at all if you can't actually use it like a map? Applies to TS/Python too.

I get the idea of wanting to iterate over them or something - but in that case maybe these should just define a `as_collection()` or `iterator()` or similar method which returns a read-only view for use in those cases.

Reply:
Have to support setting existing values (so we have to accept maps in start options) and returning this extension. Java can have overloads, so we don't have to accept a map type on start, but we do have to return this from getter or else we need to have a different name.